src/status_im2/contexts/chat/composer/link_preview/events_test.cljs

```diff
@@ -0,0 +1,212 @@
+(ns status-im2.contexts.chat.composer.link-preview.events-test
+  (:require [status-im2.contexts.chat.composer.link-preview.events :as events]
+            [cljs.test :refer [is deftest testing]]))
+
+(def url-github "https://github.com")
+(def url-gitlab "https://gitlab.com")
+(def preview-github {:url url-github :thumbnail nil})
+(def preview-gitlab {:url url-gitlab :thumbnail nil})
+(def request-id "fb973113-86f3-49c0-b750-5734a2e77a90")
+
+(deftest unfurl-urls
+  (testing "clear up state when text is empty"
+    (let [cofx                       {:db {:chat/link-previews {:unfurled   {}
+                                                                :cache      {}
+                                                                :request-id "123"}}}
+          {:keys [db json-rpc/call]} (events/unfurl-urls cofx " ")]
+      (is (= {:chat/link-previews {:cache {}}}
+             db))
+      (is (nil? call))))
+
+  (testing "fetches parsed URLs"
+    (let [cofx                       {:db {:chat/link-previews {:unfurled   {}
+                                                                :cache      {}
+                                                                :request-id "123"}}}
+          {:keys [db json-rpc/call]} (events/unfurl-urls cofx url-github)]
+      (is (nil? db))
+      (is (= {:params [url-github]}
+             (select-keys (first call) [:params]))))))
+
+(defn assert-request-id
+  [db]
+  (is (string? (get-in db [:chat/link-previews :request-id]))))
+
+(defn assert-rpc-params
+  [expected rpc-call]
+  (is (= expected
+         (select-keys (first rpc-call) [:params]))))
+
+(deftest unfurl-parsed-urls-test
+  (testing "empty state with 2 stale URLs"
+    (let [cofx                       {:db {:chat/link-previews
+                                           {:cache    {}
+                                            :unfurled []}}}
+          {:keys [db json-rpc/call]} (events/unfurl-parsed-urls cofx [url-github url-gitlab])]
+      (assert-request-id db)
+      (assert-rpc-params {:params [[url-github url-gitlab]]} call)
```

Comment:
Again, if this were me, I'd try to assert the map that is returned from the `unfurl-parsed-urls` fn, while minimizing the use of helper functions like `assert-rpc-params`:
```clojure
{:json-rpc/call
[{:method     "wakuext_unfurlURLs"
   :params     [new-urls]
   :on-success #(rf/dispatch [:link-preview/unfurl-parsed-urls-success request-id %])
   :on-error   #(rf/dispatch [:link-preview/unfurl-parsed-urls-error request-id %])}]}
```
I had a good experience using a `dispatcher` function for the dispatches and using `with-redef` to assert the params, like [this](https://github.com/status-im/status-mobile/blob/develop/src/status_im2/contexts/add_new_contact/events_test.cljs#L85). You may remember that code, you commented on it in the PR.

I'm not 100% sure there wouldn't be problems with following this approach here, just mentioning here in case you find it helpful. Again, no worries about adopting any of this feedback; it sounded like you wanted some eyes on the tests. I like the approach of testing event handlers with pure functions too! Pure Functions FTW! Speaking of which, I wonder if that random-uuid should be a coeffect?

Reply:
> Speaking of which, I wonder if that random-uuid should be a coeffect

Absolutely! What a perfect candidate for a coeffect.

> just mentioning here in case you find it helpful. Again, no worries about adopting any of this feedback

Thank you for carefully taking a look at the tests. I'll see about applying your suggestions.

> I had a good experience using a dispatcher function for the dispatches and using with-redef to assert the params, like [this](https://github.com/status-im/status-mobile/blob/develop/src/status_im2/contexts/add_new_contact/events_test.cljs#L85). You may remember that code, you commented on it in the PR.

Yes, I remember that approach with a dispatcher and with-redef, quite flexible! This time around I experimented testing the event handlers in isolation as you noticed, so the `on-success` and `on-error` functions are not covered by tests. I was thinking of using a combination of the "pure approach" for the majority of the tests and one or two more integration like tests using `day8.re-frame.test/run-test-sync`, all in the same namespace, but in the end I didn't write a happy path test using `run-test-sync`.

Reply:
@erikseppanen, I successfully applied your suggestions to the tests and they surely simplified things. Especially the idea to mock the `request-id` generator. Thanks a lot!

I went for a slightly different direction, I'll try to explain:

Almost every event handler test that we write needs at some point to compare the `json-rpc/call` effect with something. The problem is a known one, and I mentioned a couple times, the effect uses `on-success` and `on-error` callbacks, which can't be easily matched against (calling the event handler always generates a new anonymous on-success/on-error function). The correct way was if the effect was following the convention to represent on-success and on-error as vectors (re-frame event vectors). That way, it would be possible to assert in tests against the full `json-rpc/call` effect.

Since that's not currently possible, I wrote this little helper function (the `if` is really necessary btw):

```clojure
(defn remove-rpc-callbacks
  [rpc-effect]
  (if (get-in rpc-effect [:json-rpc/call 0])
    (update-in rpc-effect [:json-rpc/call 0] dissoc :on-success :on-error)
    rpc-effect))
```

And then in the tests I can do something like this, which, as you said, is simpler and clearer than destructuring and creating custom assertion functions.

```clojure
(is (= {:db            {:chat/link-previews
                        {:request-id request-id
                         :cache      cache
                         :unfurled   [preview-github
                                      {:url url-gitlab :loading? true}]}}
        :json-rpc/call [{:method "wakuext_unfurlURLs"
                         :params [[url-gitlab]]}]}
       (remove-rpc-callbacks
        (events/unfurl-parsed-urls cofx [url-github url-gitlab]))))
```

I would love to have more time at hand to refactor the effect to use pure data, so that the majority of our event handlers can be easily tested using a simple equality check.